### ISSUE TITLE
Memory Error During Video Analysis

### ISSUE THREAD

**Issue by mfkeles** [DeepLabCut/DeepLabCut#1468]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

Memory fills up during video analysis. This issue happens using multi-animal model and attempting to analyze single video file containing 1.7 million frames at 1100x800 pixels. Following setups are tested, all failed to complete analysis due to memory issues :

- TF2.5 Docker Container from Nvidia with DeepLabCut 2.2, dlcrnet
- Windows 10 with DLC 2.2, dlcrnet
- Ubuntu 20.04 with DLC 2.2, dlcrnet

Video was successfully analyzed with a single animal model (resnet-50), albeit with significant memory usage (30gb) using DLC 2.2. 

### Operating System

Tested the analysis in Windows 10, Ubuntu 20.04, and TF2.5 Nvidia Docker Container.

### DeepLabCut version

dlc version: 2.2

### DeepLabCut mode

multi animal

### Device type

gpu - RTX3090

### Steps To Reproduce

1. Open DEEPLABCUT environment, created with the provided conda file.
2. Use dlcrnet and attempt to analyze video. 
3. Memory slowly fills up (system has 64gb of memory). 
4. Error thrown related to not being able to allocate space for an array.

### Relevant log output

```shell
MemoryError                               Traceback (most recent call last)
<ipython-input-7-972df93ca544> in <module>
----> 1 deeplabcut.analyze_videos(config,[vid_path],gputouse=2)

~\Desktop\MK\Git\DeepLabCut\deeplabcut\pose_estimation_tensorflow\predict_videos.py in analyze_videos(config, videos, videotype, shuffle, trainingsetindex, gputouse, save_as_csv, destfolder, batchsize, cropping, TFGPUinference, dynamic, modelprefix, robust_nframes, allow_growth)
    286             )
    287             for video in Videos:
--> 288                 AnalyzeMultiAnimalVideo(
    289                     video,
    290                     DLCscorer,

~\Desktop\MK\Git\DeepLabCut\deeplabcut\pose_estimation_tensorflow\predict_multianimal.py in AnalyzeMultiAnimalVideo(video, DLCscorer, trainFraction, cfg, dlc_cfg, sess, inputs, outputs, destfolder, robust_nframes)
     77         print("Starting to extract posture")
     78         if int(dlc_cfg["batch_size"]) > 1:
---> 79             PredicteData, nframes = GetPoseandCostsF(
     80                 cfg,
     81                 dlc_cfg,

~\Desktop\MK\Git\DeepLabCut\deeplabcut\pose_estimation_tensorflow\predict_multianimal.py in GetPoseandCostsF(cfg, dlc_cfg, sess, inputs, outputs, cap, nframes, batchsize)
    145             inds.append(counter)
    146             if batch_ind == batchsize - 1:
--> 147                 D = predict.predict_batched_peaks_and_costs(
    148                     dlc_cfg, frames, sess, inputs, outputs,
    149                 )

~\Desktop\MK\Git\DeepLabCut\deeplabcut\pose_estimation_tensorflow\core\predict_multianimal.py in predict_batched_peaks_and_costs(pose_cfg, images_batch, sess, inputs, outputs, peaks_gt, n_points, n_decimals)
    208     n_decimals=3,
    209 ):
--> 210     scmaps, locrefs, pafs, peaks = sess.run(
    211         outputs, feed_dict={inputs: images_batch}
    212     )

~\anaconda3\envs\DEEPLABCUT\lib\site-packages\tensorflow\python\client\session.py in run(self, fetches, feed_dict, options, run_metadata)
    965
    966     try:
--> 967       result = self._run(None, fetches, feed_dict, options_ptr,
    968                          run_metadata_ptr)
    969       if run_metadata:

~\anaconda3\envs\DEEPLABCUT\lib\site-packages\tensorflow\python\client\session.py in _run(self, handle, fetches, feed_dict, options, run_metadata)
   1158             feed_handles[subfeed_t.ref()] = subfeed_val
   1159           else:
-> 1160             np_val = np.asarray(subfeed_val, dtype=subfeed_dtype)
   1161
   1162           if (not is_tensor_handle_feed and

~\anaconda3\envs\DEEPLABCUT\lib\site-packages\numpy\core\_asarray.py in asarray(a, dtype, order)
     81
     82     """
---> 83     return array(a, dtype, copy=False, order=order)
     84
     85

MemoryError: Unable to allocate 40.3 MiB for an array with shape (4, 800, 1100, 3) and data type float32
```

### Anything else?

I would be happy to share my video file (14.5gb) along with the model for reproducibility. It is obviously a very large dataset and i am not sure if it is feasible to analyze this as a single file. I am aware of the built in video functions to split large files and would be happy to workaround this issue by splitting the video. That said, I am curious if there is a straightforward solution in the code that could work. 

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by jeylau** [comment]
Hey @mfkeles, that looks like a serious memory leak. Thanks for reporting it, I'll see if I can find its cause 😎

**Comment by mfkeles** [comment]
@jeylau Thank you so much!!
For what it's worth, memory issue repeated on [Colab](https://colab.research.google.com/drive/1pt0q7p8JsXKOm2cd-JGmg_Wl1VjuKej4?usp=sharing) as well.

**Comment by AlexEMG** [comment]
We think this might be an OpenCV memory leak. Could you comment on what frame it fails (and if it consistently fails at the same fraction of your video)?

**Comment by mfkeles** [comment]
It doesn't fail at the same frame (or at least the % of analyzed vid).  With 64 gb memory, it fails around 100k (~5%) frames and with lower memory systems (colab) it fails around 20k (~1%).

**Comment by mfkeles** [comment]
hey @jeylau , just checking in to see if there is anything I can help with on this! 

Thank you!!!

**Comment by jeylau** [comment]
hey @mfkeles, could you please share the ouput of `conda env export`? I can't observe that behavior with DLC 2.2.0.2 (although nothing has changed much since 2.2). I see constant memory use (i.e., fluctuations of <1%) over 15k frames of a ~1000x1000 RGB video, using either opencv or pyav (although pyav does look a bit more stable by the way). So I suspect a particular package version may cause it 🤔

**Comment by mfkeles** [comment]
hey @jeylau, hmmm, I have tried it on the [google colab ](https://colab.research.google.com/drive/1C5FVySeeyjxOGXQ6cWFdaLMUWtd1iLP0?usp=sharing) with v. 2.2.0.2 and it still runs out of memory around 20k. Here is also the conda env I use on the local machine that gave me the error:
```

channels:
  - cctbx202008
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=4.5=1_gnu
  - argon2-cffi=20.1.0=py38h497a2fe_2
  - async_generator=1.10=py_0
  - atk=2.36.0=1
  - atk-1.0=2.36.0=haf93ef1_1
  - attrs=21.2.0=pyhd8ed1ab_0
  - backcall=0.2.0=pyh9f0ad1d_0
  - backports=1.0=py_2
  - backports.functools_lru_cache=1.6.4=pyhd8ed1ab_0
  - bleach=3.3.0=pyh44b312d_0
  - blosc=1.21.0=h9c3ff4c_0
  - brotlipy=0.7.0=py38h1e0a361_1000
  - bzip2=1.0.8=h7f98852_4
  - c-ares=1.17.1=h7f98852_1
  - ca-certificates=2021.5.30=ha878542_0
  - cairo=1.16.0=hcf35c78_1003
  - certifi=2021.5.30=py38h578d9bd_0
  - cffi=1.14.1=py38h5bae8af_0
  - cryptography=3.0=py38h766eaa4_0
  - cudatoolkit=11.2.2=he111cf0_8
  - cudnn=8.2.1.32=h86fa8c9_0
  - dbus=1.13.18=hb2f20db_0
  - debugpy=1.3.0=py38h709712a_0
  - defusedxml=0.7.1=pyhd8ed1ab_0
  - entrypoints=0.3=pyhd8ed1ab_1003
  - expat=2.2.10=h9c3ff4c_0
  - ffmpeg=4.3.1=h167e202_0
  - fontconfig=2.13.1=h86ecdb6_1001
  - freetype=2.10.2=he06d7ca_0
  - fribidi=1.0.10=h516909a_0
  - gdk-pixbuf=2.38.2=h3f25603_4
  - gettext=0.19.8.1=hc5be6a0_1002
  - glib=2.65.0=h6f030ca_0
  - gmp=6.2.1=h58526e2_0
  - gnutls=3.6.13=h85f3911_1
  - gobject-introspection=1.64.1=py38h03d966d_1
  - graphite2=1.3.13=he1b5a44_1001
  - gst-plugins-base=1.14.5=h0935bb2_2
  - gstreamer=1.14.5=h36ae1b5_2
  - gtk2=2.24.32=h586f36d_1
  - harfbuzz=2.4.0=h9f30f68_3
  - hdf5=1.10.6=nompi_h6a2412b_1114
  - icu=64.2=he1b5a44_1
  - idna=2.10=pyh9f0ad1d_0
  - importlib-metadata=4.6.1=py38h578d9bd_0
  - ipykernel=6.0.1=py38hd0cf306_0
  - ipython=7.25.0=py38hd0cf306_1
  - ipython_genutils=0.2.0=py_1
  - ipywidgets=7.6.3=pyhd3deb0d_0
  - jedi=0.18.0=py38h578d9bd_2
  - jinja2=3.0.1=pyhd8ed1ab_0
  - jpeg=9d=h516909a_0
  - jsonschema=3.2.0=pyhd8ed1ab_3
  - jupyter=1.0.0=py38h578d9bd_6
  - jupyter_client=6.1.12=pyhd8ed1ab_0
  - jupyter_console=6.4.0=pyhd8ed1ab_0
  - jupyter_core=4.7.1=py38h578d9bd_0
  - jupyterlab_pygments=0.1.2=pyh9f0ad1d_0
  - jupyterlab_widgets=1.0.0=pyhd8ed1ab_1
  - krb5=1.19.1=hcc1bbae_0
  - lame=3.100=h7f98852_1001
  - ld_impl_linux-64=2.35.1=h7274673_9
  - libblas=3.9.0=9_openblas
  - libcblas=3.9.0=9_openblas
  - libcurl=7.77.0=h2574ce0_0
  - libedit=3.1.20191231=he28a2e2_2
  - libev=4.33=h516909a_1
  - libffi=3.2.1=he1b5a44_1007
  - libgcc-ng=9.3.0=h5101ec6_17
  - libgfortran-ng=9.3.0=hff62375_19
  - libgfortran5=9.3.0=hff62375_19
  - libglu=9.0.0=he1b5a44_1001
  - libgomp=9.3.0=h5101ec6_17
  - libiconv=1.15=h516909a_1006
  - liblapack=3.9.0=9_openblas
  - libnghttp2=1.43.0=h812cca2_0
  - libopenblas=0.3.15=pthreads_h8fe5266_1
  - libpng=1.6.37=hed695b0_2
  - libsodium=1.0.18=h36c2ea0_1
  - libssh2=1.9.0=ha56f1ee_6
  - libstdcxx-ng=9.3.0=hd4cf53a_17
  - libtiff=4.1.0=hc7e4089_6
  - libuuid=2.32.1=h14c3975_1000
  - libwebp-base=1.1.0=h516909a_3
  - libxcb=1.13=h14c3975_1002
  - libxml2=2.9.10=hee79883_0
  - lz4-c=1.8.3=he1b5a44_1001
  - lzo=2.10=h516909a_1000
  - markupsafe=2.0.1=py38h497a2fe_0
  - matplotlib-inline=0.1.2=pyhd8ed1ab_2
  - mistune=0.8.4=py38h497a2fe_1004
  - mock=4.0.3=py38h578d9bd_1
  - nb_conda=2.2.1=py38h578d9bd_4
  - nb_conda_kernels=2.3.1=py38h578d9bd_0
  - nbclient=0.5.3=pyhd8ed1ab_0
  - nbconvert=6.1.0=py38h578d9bd_0
  - nbformat=5.1.3=pyhd8ed1ab_0
  - ncurses=6.2=he6710b0_1
  - nest-asyncio=1.5.1=pyhd8ed1ab_0
  - nettle=3.6=he412f7d_0
  - notebook=6.4.0=pyha770c72_0
  - numexpr=2.7.3=py38h51da96c_0
  - openh264=2.1.1=h780b84a_0
  - openssl=1.1.1k=h7f98852_0
  - packaging=21.0=pyhd8ed1ab_0
  - pandoc=2.14.0.3=h7f98852_0
  - pandocfilters=1.4.2=py_1
  - pango=1.42.4=h7062337_4
  - parso=0.8.2=pyhd8ed1ab_0
  - pathlib2=2.3.5=py38h32f6830_1
  - pcre=8.44=he1b5a44_0
  - pexpect=4.8.0=pyh9f0ad1d_2
  - pickleshare=0.7.5=py_1003
  - pip=21.1.3=py38h06a4308_0
  - pixman=0.38.0=h516909a_1003
  - prometheus_client=0.11.0=pyhd8ed1ab_0
  - prompt-toolkit=3.0.19=pyha770c72_0
  - prompt_toolkit=3.0.19=hd8ed1ab_0
  - pthread-stubs=0.4=h14c3975_1001
  - ptyprocess=0.7.0=pyhd3deb0d_0
  - pycparser=2.20=pyh9f0ad1d_2
  - pygments=2.9.0=pyhd8ed1ab_0
  - pyopenssl=19.1.0=py_1
  - pyparsing=2.4.7=pyh9f0ad1d_0
  - pypubsub=4.0.3=py_0
  - pyqt=5.9.2=py38h05f1152_4
  - pyrsistent=0.17.3=py38h497a2fe_2
  - pysocks=1.7.1=py38h32f6830_1
  - pytables=3.6.1=py38hc386592_3
  - python=3.8.5=h4d41432_2_cpython
  - python-dateutil=2.8.1=py_0
  - python_abi=3.8=2_cp38
  - pyzmq=22.1.0=py38h2035c66_0
  - qt=5.9.7=h0c104cb_3
  - qtconsole=5.1.1=pyhd8ed1ab_0
  - qtpy=1.9.0=py_0
  - readline=8.1=h27cfd23_0
  - send2trash=1.7.1=pyhd8ed1ab_0
  - setuptools=52.0.0=py38h06a4308_0
  - sip=4.19.13=py38he6710b0_0
  - six=1.15.0=pyh9f0ad1d_0
  - sqlite=3.36.0=hc218d9a_0
  - terminado=0.10.1=py38h578d9bd_0
  - testpath=0.5.0=pyhd8ed1ab_0
  - tk=8.6.10=hbc83047_0
  - tornado=6.1=py38h497a2fe_1
  - traitlets=5.0.5=py_0
  - wcwidth=0.2.5=pyh9f0ad1d_2
  - webencodings=0.5.1=py_1
  - wheel=0.36.2=pyhd3eb1b0_0
  - widgetsnbextension=3.5.1=py38h578d9bd_4
  - wxpython=4.0.7.post2=py38hac52f33_3
  - x264=1!152.20180806=h14c3975_0
  - xorg-kbproto=1.0.7=h14c3975_1002
  - xorg-libice=1.0.10=h516909a_0
  - xorg-libsm=1.2.3=h84519dc_1000
  - xorg-libx11=1.6.11=h516909a_0
  - xorg-libxau=1.0.9=h14c3975_0
  - xorg-libxdmcp=1.1.3=h516909a_0
  - xorg-libxext=1.3.4=h516909a_0
  - xorg-libxrender=0.9.10=h516909a_1002
  - xorg-renderproto=0.11.1=h14c3975_1002
  - xorg-xextproto=7.3.0=h14c3975_1002
  - xorg-xproto=7.0.31=h14c3975_1007
  - xz=5.2.5=h7b6447c_0
  - zeromq=4.3.4=h9c3ff4c_0
  - zipp=3.5.0=pyhd8ed1ab_0
  - zlib=1.2.11=h7b6447c_3
  - zstd=1.4.4=h3b9ef0a_2
  - pip:
    - absl-py==0.13.0
    - astunparse==1.6.3
    - bayesian-optimization==1.2.0
    - cachetools==4.2.2
    - chardet==4.0.0
    - click==8.0.1
    - cycler==0.10.0
    - cython==0.29.23
    - decorator==4.4.2
    - deeplabcut==2.2.0.2
    - filterpy==1.4.5
    - flatbuffers==1.12
    - gast==0.4.0
    - google-auth==1.32.1
    - google-auth-oauthlib==0.4.4
    - google-pasta==0.2.0
    - grpcio==1.34.1
    - h5py==3.1.0
    - imageio==2.9.0
    - imageio-ffmpeg==0.4.4
    - imgaug==0.4.0
    - intel-openmp==2021.3.0
    - ipython-genutils==0.2.0
    - joblib==1.0.1
    - keras-nightly==2.5.0.dev2021032900
    - keras-preprocessing==1.1.2
    - kiwisolver==1.3.1
    - llvmlite==0.34.0
    - markdown==3.3.4
    - matplotlib==3.1.3
    - moviepy==1.0.1
    - msgpack==1.0.2
    - msgpack-numpy==0.4.7.1
    - networkx==2.5.1
    - numba==0.51.1
    - numpy==1.19.5
    - oauthlib==3.1.1
    - opencv-python==4.5.2.54
    - opencv-python-headless==3.4.9.33
    - opt-einsum==3.3.0
    - pandas==1.3.0
    - patsy==0.5.1
    - pillow==8.3.0
    - proglog==0.1.9
    - protobuf==3.17.3
    - psutil==5.8.0
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pytz==2021.1
    - pywavelets==1.1.1
    - pyyaml==5.4.1
    - requests==2.25.1
    - requests-oauthlib==1.3.0
    - rsa==4.7.2
    - ruamel-yaml==0.17.10
    - ruamel-yaml-clib==0.2.6
    - scikit-image==0.18.2
    - scikit-learn==0.24.2
    - scipy==1.7.0
    - shapely==1.7.1
    - statsmodels==0.12.2
    - tabulate==0.8.9
    - tensorboard==2.5.0
    - tensorboard-data-server==0.6.1
    - tensorboard-plugin-wit==1.8.0
    - tensorflow==2.5.0
    - tensorflow-estimator==2.5.0
    - tensorpack==0.9.8
    - termcolor==1.1.0
    - tf-slim==1.1.0
    - threadpoolctl==2.1.0
    - tifffile==2021.7.2
    - tqdm==4.61.2
    - typing-extensions==3.7.4.3
    - urllib3==1.26.6
    - werkzeug==2.0.1
    - wrapt==1.12.1
```

**Comment by mfkeles** [comment]
I can confirm that with `    - opencv-python==4.5.3.56
    - opencv-python-headless==4.5.3.56` I still have the constant increase in memory.

**Comment by MMathisLab** [comment]
@mfkeles could you try the new DLC docker? `pip install deeplabcut-docker` also looking at the conda, it seems you likely upgraded DLC in an old conda, could you try also a fresh conda DEEPLABCUT.yaml?

**Comment by jeylau** [comment]
@mfkeles, I'm not observing anything strange using your environment. What's very odd though is that purely decoding video frames with opencv ran at ~450fps with yours, down from 950+ with mine (with no clear diff between the two).
```
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=4.5=1_gnu
  - backcall=0.2.0=pyhd3eb1b0_0
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2021.10.8=ha878542_0
  - certifi=2021.10.8=py38h578d9bd_0
  - cudatoolkit=11.3.1=h2bc3f7f_2
  - cudnn=8.2.1=cuda11.3_0
  - debugpy=1.4.1=py38h295c915_0
  - entrypoints=0.3=py38_0
  - ffmpeg=4.3.1=hca11adc_2
  - freetype=2.10.4=h5ab3b9f_0
  - gmp=6.2.1=h2531618_2
  - gnutls=3.6.15=he1e5248_0
  - ipykernel=6.4.1=py38he5a9106_0
  - ipython=7.27.0=py38hb070fc8_0
  - ipython_genutils=0.2.0=pyhd3eb1b0_1
  - jedi=0.18.0=py38h06a4308_1
  - jupyter_client=7.0.1=pyhd3eb1b0_0
  - jupyter_core=4.8.1=py38h06a4308_0
  - lame=3.100=h7b6447c_0
  - ld_impl_linux-64=2.35.1=h7274673_9
  - libffi=3.3=he6710b0_2
  - libgcc-ng=9.3.0=h5101ec6_17
  - libgomp=9.3.0=h5101ec6_17
  - libidn2=2.3.2=h7f8727e_0
  - libopus=1.3.1=h7b6447c_0
  - libpng=1.6.37=hbc83047_0
  - libsodium=1.0.18=h7b6447c_0
  - libstdcxx-ng=9.3.0=hd4cf53a_17
  - libtasn1=4.16.0=h27cfd23_0
  - libunistring=0.9.10=h27cfd23_0
  - libvpx=1.7.0=h439df22_0
  - matplotlib-inline=0.1.2=pyhd3eb1b0_2
  - ncurses=6.2=he6710b0_1
  - nest-asyncio=1.5.1=pyhd3eb1b0_0
  - nettle=3.7.3=hbbd107a_1
  - openh264=2.1.1=h780b84a_0
  - openssl=1.1.1l=h7f8727e_0
  - parso=0.8.2=pyhd3eb1b0_0
  - pexpect=4.8.0=pyhd3eb1b0_3
  - pickleshare=0.7.5=pyhd3eb1b0_1003
  - pip=21.2.4=py38h06a4308_0
  - prompt-toolkit=3.0.20=pyhd3eb1b0_0
  - ptyprocess=0.7.0=pyhd3eb1b0_2
  - pygments=2.10.0=pyhd3eb1b0_0
  - python=3.8.12=h12debd9_0
  - python-dateutil=2.8.2=pyhd3eb1b0_0
  - python_abi=3.8=2_cp38
  - readline=8.1=h27cfd23_0
  - setuptools=58.0.4=py38h06a4308_0
  - sqlite=3.36.0=hc218d9a_0
  - tk=8.6.11=h1ccaba5_0
  - tornado=6.1=py38h27cfd23_0
  - traitlets=5.1.0=pyhd3eb1b0_0
  - wcwidth=0.2.5=pyhd3eb1b0_0
  - wheel=0.37.0=pyhd3eb1b0_1
  - x264=1!161.3030=h7f98852_1
  - xz=5.2.5=h7b6447c_0
  - zeromq=4.3.4=h2531618_0
  - zlib=1.2.11=h7b6447c_3
  - pip:
    - absl-py==0.14.1
    - astunparse==1.6.3
    - attrs==21.2.0
    - av==8.0.3
    - cachetools==4.2.4
    - charset-normalizer==2.0.7
    - clang==5.0
    - coverage==6.0.2
    - cycler==0.10.0
    - decorator==4.4.2
    - filterpy==1.4.5
    - flatbuffers==1.12
    - gast==0.4.0
    - google-auth==2.3.0
    - google-auth-oauthlib==0.4.6
    - google-pasta==0.2.0
    - grpcio==1.41.0
    - h5py==3.1.0
    - idna==3.3
    - imageio==2.9.0
    - imageio-ffmpeg==0.4.5
    - imgaug==0.4.0
    - iniconfig==1.1.1
    - joblib==1.1.0
    - keras==2.6.0
    - keras-preprocessing==1.1.2
    - kiwisolver==1.3.2
    - llvmlite==0.37.0
    - markdown==3.3.4
    - matplotlib==3.4.3
    - moviepy==1.0.3
    - msgpack==1.0.2
    - msgpack-numpy==0.4.7.1
    - networkx==2.6.3
    - numba==0.54.1
    - numexpr==2.7.3
    - numpy==1.19.5
    - oauthlib==3.1.1
    - opencv-python==4.5.3.56
    - opencv-python-headless==4.5.3.56
    - opt-einsum==3.3.0
    - packaging==21.0
    - pandas==1.3.4
    - patsy==0.5.2
    - pillow==8.4.0
    - pluggy==1.0.0
    - proglog==0.1.9
    - protobuf==3.18.1
    - psutil==5.8.0
    - py==1.10.0
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pyparsing==2.4.7
    - pytest==6.2.5
    - pytest-cov==3.0.0
    - pytest-cover==3.0.0
    - pytest-coverage==0.0
    - pytz==2021.3
    - pywavelets==1.1.1
    - pyyaml==6.0
    - pyzmq==22.3.0
    - requests==2.26.0
    - requests-oauthlib==1.3.0
    - rsa==4.7.2
    - ruamel-yaml==0.17.16
    - ruamel-yaml-clib==0.2.6
    - scikit-image==0.18.3
    - scikit-learn==1.0
    - scipy==1.7.1
    - shapely==1.7.1
    - six==1.15.0
    - statsmodels==0.13.0
    - tables==3.6.1
    - tabulate==0.8.9
    - tensorboard==2.7.0
    - tensorboard-data-server==0.6.1
    - tensorboard-plugin-wit==1.8.0
    - tensorflow==2.6.0
    - tensorflow-estimator==2.6.0
    - tensorpack==0.11
    - termcolor==1.1.0
    - tf-slim==1.1.0
    - threadpoolctl==3.0.0
    - tifffile==2021.10.12
    - toml==0.10.2
    - tomli==1.2.1
    - tqdm==4.62.3
    - typing-extensions==3.7.4.3
    - urllib3==1.26.7
    - werkzeug==2.0.2
    - wrapt==1.12.1
```
I don't know if you can spot something, but maybe that diff in speed is indirectly related to your issue 🤔

**Comment by jeylau** [comment]
@mfkeles, I'm not observing anything strange using your environment. What's very odd though is that purely decoding video frames with opencv ran at ~450fps with yours, down from 950+ with mine (with no clear diff between the two).
```
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=4.5=1_gnu
  - backcall=0.2.0=pyhd3eb1b0_0
  - bzip2=1.0.8=h7b6447c_0
  - ca-certificates=2021.10.8=ha878542_0
  - certifi=2021.10.8=py38h578d9bd_0
  - cudatoolkit=11.3.1=h2bc3f7f_2
  - cudnn=8.2.1=cuda11.3_0
  - debugpy=1.4.1=py38h295c915_0
  - entrypoints=0.3=py38_0
  - ffmpeg=4.3.1=hca11adc_2
  - freetype=2.10.4=h5ab3b9f_0
  - gmp=6.2.1=h2531618_2
  - gnutls=3.6.15=he1e5248_0
  - ipykernel=6.4.1=py38he5a9106_0
  - ipython=7.27.0=py38hb070fc8_0
  - ipython_genutils=0.2.0=pyhd3eb1b0_1
  - jedi=0.18.0=py38h06a4308_1
  - jupyter_client=7.0.1=pyhd3eb1b0_0
  - jupyter_core=4.8.1=py38h06a4308_0
  - lame=3.100=h7b6447c_0
  - ld_impl_linux-64=2.35.1=h7274673_9
  - libffi=3.3=he6710b0_2
  - libgcc-ng=9.3.0=h5101ec6_17
  - libgomp=9.3.0=h5101ec6_17
  - libidn2=2.3.2=h7f8727e_0
  - libopus=1.3.1=h7b6447c_0
  - libpng=1.6.37=hbc83047_0
  - libsodium=1.0.18=h7b6447c_0
  - libstdcxx-ng=9.3.0=hd4cf53a_17
  - libtasn1=4.16.0=h27cfd23_0
  - libunistring=0.9.10=h27cfd23_0
  - libvpx=1.7.0=h439df22_0
  - matplotlib-inline=0.1.2=pyhd3eb1b0_2
  - ncurses=6.2=he6710b0_1
  - nest-asyncio=1.5.1=pyhd3eb1b0_0
  - nettle=3.7.3=hbbd107a_1
  - openh264=2.1.1=h780b84a_0
  - openssl=1.1.1l=h7f8727e_0
  - parso=0.8.2=pyhd3eb1b0_0
  - pexpect=4.8.0=pyhd3eb1b0_3
  - pickleshare=0.7.5=pyhd3eb1b0_1003
  - pip=21.2.4=py38h06a4308_0
  - prompt-toolkit=3.0.20=pyhd3eb1b0_0
  - ptyprocess=0.7.0=pyhd3eb1b0_2
  - pygments=2.10.0=pyhd3eb1b0_0
  - python=3.8.12=h12debd9_0
  - python-dateutil=2.8.2=pyhd3eb1b0_0
  - python_abi=3.8=2_cp38
  - readline=8.1=h27cfd23_0
  - setuptools=58.0.4=py38h06a4308_0
  - sqlite=3.36.0=hc218d9a_0
  - tk=8.6.11=h1ccaba5_0
  - tornado=6.1=py38h27cfd23_0
  - traitlets=5.1.0=pyhd3eb1b0_0
  - wcwidth=0.2.5=pyhd3eb1b0_0
  - wheel=0.37.0=pyhd3eb1b0_1
  - x264=1!161.3030=h7f98852_1
  - xz=5.2.5=h7b6447c_0
  - zeromq=4.3.4=h2531618_0
  - zlib=1.2.11=h7b6447c_3
  - pip:
    - absl-py==0.14.1
    - astunparse==1.6.3
    - attrs==21.2.0
    - av==8.0.3
    - cachetools==4.2.4
    - charset-normalizer==2.0.7
    - clang==5.0
    - coverage==6.0.2
    - cycler==0.10.0
    - decorator==4.4.2
    - filterpy==1.4.5
    - flatbuffers==1.12
    - gast==0.4.0
    - google-auth==2.3.0
    - google-auth-oauthlib==0.4.6
    - google-pasta==0.2.0
    - grpcio==1.41.0
    - h5py==3.1.0
    - idna==3.3
    - imageio==2.9.0
    - imageio-ffmpeg==0.4.5
    - imgaug==0.4.0
    - iniconfig==1.1.1
    - joblib==1.1.0
    - keras==2.6.0
    - keras-preprocessing==1.1.2
    - kiwisolver==1.3.2
    - llvmlite==0.37.0
    - markdown==3.3.4
    - matplotlib==3.4.3
    - moviepy==1.0.3
    - msgpack==1.0.2
    - msgpack-numpy==0.4.7.1
    - networkx==2.6.3
    - numba==0.54.1
    - numexpr==2.7.3
    - numpy==1.19.5
    - oauthlib==3.1.1
    - opencv-python==4.5.3.56
    - opencv-python-headless==4.5.3.56
    - opt-einsum==3.3.0
    - packaging==21.0
    - pandas==1.3.4
    - patsy==0.5.2
    - pillow==8.4.0
    - pluggy==1.0.0
    - proglog==0.1.9
    - protobuf==3.18.1
    - psutil==5.8.0
    - py==1.10.0
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pyparsing==2.4.7
    - pytest==6.2.5
    - pytest-cov==3.0.0
    - pytest-cover==3.0.0
    - pytest-coverage==0.0
    - pytz==2021.3
    - pywavelets==1.1.1
    - pyyaml==6.0
    - pyzmq==22.3.0
    - requests==2.26.0
    - requests-oauthlib==1.3.0
    - rsa==4.7.2
    - ruamel-yaml==0.17.16
    - ruamel-yaml-clib==0.2.6
    - scikit-image==0.18.3
    - scikit-learn==1.0
    - scipy==1.7.1
    - shapely==1.7.1
    - six==1.15.0
    - statsmodels==0.13.0
    - tables==3.6.1
    - tabulate==0.8.9
    - tensorboard==2.7.0
    - tensorboard-data-server==0.6.1
    - tensorboard-plugin-wit==1.8.0
    - tensorflow==2.6.0
    - tensorflow-estimator==2.6.0
    - tensorpack==0.11
    - termcolor==1.1.0
    - tf-slim==1.1.0
    - threadpoolctl==3.0.0
    - tifffile==2021.10.12
    - toml==0.10.2
    - tomli==1.2.1
    - tqdm==4.62.3
    - typing-extensions==3.7.4.3
    - urllib3==1.26.7
    - werkzeug==2.0.2
    - wrapt==1.12.1
```
I don't know if you can spot something, but maybe that diff in speed indirectly relates to your issue 🤔

**Comment by mfkeles** [comment]
@MMathisLab @jeylau thank you for the suggestions! So I tried deeplabcut-docker, fresh installation from DEEPLABCUT.yaml and @jeylau's environment without any improvement. Memory usage steadily increases and never stops even after 100k frames. It eventually fills up all the available ram.  😭 😭 😭 😭 

I would be more than happy to share my project folder along with the video if it might help replicating this issue on your end! Please let me know.

**Comment by MorterzaAbbaszadeh** [comment]
Hey:)
I have a similar but a bit wierder problem. 
I am using 2.1.9, conda virtual env on a local ubuntu machine running Python 3.7.9. After a year of working fine, begining from last week it stopped producing h5 and CSV files after analysis, but only when analyzing videos with more than 10k frames, runs perfectly fine for lower than 10K frames, or the same videos cut for lower than 10K frames.
 Analysis runs successfully. there is no error. it just doesnt produce any CSV or h5 files afterward for more than 10k frames.
Here is the env:
```
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_tflow_select             2.1.0                       gpu  
absl-py                   0.11.0           py37h06a4308_0  
argon2-cffi               20.1.0           py37h7b6447c_1  
astor                     0.8.1                    py37_0  
async_generator           1.10             py37h28b3542_0  
attrs                     20.3.0             pyhd3eb1b0_0  
backcall                  0.2.0                      py_0  
bayesian-optimization     1.2.0                    pypi_0    pypi
blas                      1.0                         mkl  
bleach                    3.2.1                      py_0  
c-ares                    1.17.1               h27cfd23_0  
ca-certificates           2020.10.14                    0  
cairo                     1.14.12              h8948797_3  
certifi                   2020.11.8        py37h06a4308_0  
cffi                      1.14.4           py37h261ae71_0  
chardet                   3.0.4                    pypi_0    pypi
click                     7.1.2                    pypi_0    pypi
cudatoolkit               10.0.130                      0  
cudnn                     7.6.5                cuda10.0_0  
cupti                     10.0.130                      0  
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.21                  pypi_0    pypi
dbus                      1.13.18              hb2f20db_0  
decorator                 4.4.2                      py_0  
deeplabcut                2.1.9                    pypi_0    pypi
defusedxml                0.6.0                      py_0  
easydict                  1.9                      pypi_0    pypi
entrypoints               0.3                      py37_0  
expat                     2.2.10               he6710b0_2  
filterpy                  1.4.5                    pypi_0    pypi
fontconfig                2.13.0               h9420a91_0  
freetype                  2.10.4               h5ab3b9f_0  
fribidi                   1.0.10               h7b6447c_0  
gast                      0.4.0                      py_0  
geos                      3.8.0                he6710b0_0  
gettext                   0.19.8.1             h9b4dc7a_1  
glib                      2.66.1               h92f7085_0  
graphite2                 1.3.14               h23475e2_0  
grpcio                    1.31.0           py37hf8bcb03_0  
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb31296c_0  
h5py                      2.10.0           py37hd6299e0_1  
harfbuzz                  2.4.0                hca77d97_1  
hdf5                      1.10.6               hb1b8bf9_0  
icu                       58.2                 he6710b0_3  
idna                      2.10                     pypi_0    pypi
imageio                   2.9.0                    pypi_0    pypi
imageio-ffmpeg            0.4.2                    pypi_0    pypi
imgaug                    0.4.0                    pypi_0    pypi
importlib-metadata        2.0.0                      py_1  
importlib_metadata        2.0.0                         1  
intel-openmp              2021.1.1                 pypi_0    pypi
ipykernel                 5.3.4            py37h5ca1d4c_0  
ipython                   7.19.0           py37hb070fc8_0  
ipython_genutils          0.2.0              pyhd3eb1b0_1  
ipywidgets                7.5.1                      py_1  
jedi                      0.17.2           py37h06a4308_1  
jinja2                    2.11.2                     py_0  
joblib                    0.17.0                   pypi_0    pypi
jpeg                      9b                   h024ee3a_2  
jsonschema                3.2.0                      py_2  
jupyter                   1.0.0                    py37_7  
jupyter_client            6.1.7                      py_0  
jupyter_console           6.2.0                      py_0  
jupyter_core              4.7.0            py37h06a4308_0  
jupyterlab_pygments       0.1.2                      py_0  
keras-applications        1.0.8                      py_1  
keras-preprocessing       1.1.0                      py_1  
kiwisolver                1.3.1                    pypi_0    pypi
ld_impl_linux-64          2.33.1               h53a641e_7  
libedit                   3.1.20191231         h14c3975_1  
libffi                    3.3                  he6710b0_2  
libgcc-ng                 9.1.0                hdf63c60_0  
libgfortran-ng            7.3.0                hdf63c60_0  
libglu                    9.0.0                hf484d3e_1  
libpng                    1.6.37               hbc83047_0  
libprotobuf               3.13.0.1             hd408876_0  
libsodium                 1.0.18               h7b6447c_0  
libstdcxx-ng              9.1.0                hdf63c60_0  
libuuid                   1.0.3                h1bed415_2  
libxcb                    1.14                 h7b6447c_0  
libxml2                   2.9.10               hb55368b_3  
llvmlite                  0.34.0                   pypi_0    pypi
markdown                  3.3.3            py37h06a4308_0  
markupsafe                1.1.1            py37h14c3975_1  
matplotlib                3.1.3                    pypi_0    pypi
mistune                   0.8.4           py37h14c3975_1001  
mkl                       2020.2                      256  
mkl-service               2.3.0            py37he8ac12f_0  
mkl_fft                   1.2.0            py37h23d657b_0  
mkl_random                1.1.1            py37h0573a6f_0  
mock                      4.0.2                      py_0  
moviepy                   1.0.1                    pypi_0    pypi
msgpack                   1.0.0                    pypi_0    pypi
msgpack-numpy             0.4.7.1                  pypi_0    pypi
nb_conda                  2.2.1                    py37_0  
nb_conda_kernels          2.3.1            py37h06a4308_0  
nbclient                  0.5.1                      py_0  
nbconvert                 6.0.7                    py37_0  
nbformat                  5.0.8                      py_0  
ncurses                   6.2                  he6710b0_1  
nest-asyncio              1.4.3              pyhd3eb1b0_0  
networkx                  2.5                      pypi_0    pypi
notebook                  6.1.4                    py37_0  
numba                     0.51.1                   pypi_0    pypi
numexpr                   2.7.1                    pypi_0    pypi
numpy                     1.16.4                   pypi_0    pypi
opencv-python             4.4.0.46                 pypi_0    pypi
opencv-python-headless    4.4.0.46                 pypi_0    pypi
openssl                   1.1.1h               h7b6447c_0  
packaging                 20.7               pyhd3eb1b0_0  
pandas                    1.1.4                    pypi_0    pypi
pandoc                    2.11                 hb0f4dca_0  
pandocfilters             1.4.3            py37h06a4308_1  
pango                     1.45.3               hd140c19_0  
parso                     0.7.0                      py_0  
patsy                     0.5.1                    pypi_0    pypi
pcre                      8.44                 he6710b0_0  
pexpect                   4.8.0              pyhd3eb1b0_3  
pickleshare               0.7.5           pyhd3eb1b0_1003  
pillow                    8.0.1                    pypi_0    pypi
pip                       20.3             py37h06a4308_0  
pixman                    0.40.0               h7b6447c_0  
proglog                   0.1.9                    pypi_0    pypi
prometheus_client         0.9.0              pyhd3eb1b0_0  
prompt-toolkit            3.0.8                      py_0  
prompt_toolkit            3.0.8                         0  
protobuf                  3.13.0.1         py37he6710b0_1  
psutil                    5.7.3                    pypi_0    pypi
ptyprocess                0.6.0              pyhd3eb1b0_2  
pycparser                 2.20                       py_2  
pygments                  2.7.2              pyhd3eb1b0_0  
pyparsing                 2.4.7                      py_0  
pyqt                      5.9.2            py37h05f1152_2  
pyrsistent                0.17.3           py37h7b6447c_0  
python                    3.7.9                h7579374_0  
python-dateutil           2.8.1                      py_0  
pytz                      2020.4                   pypi_0    pypi
pywavelets                1.1.1                    pypi_0    pypi
pyyaml                    5.3.1                    pypi_0    pypi
pyzmq                     20.0.0           py37h2531618_1  
qt                        5.9.7                h5867ecd_1  
qtconsole                 4.7.7                      py_0  
qtpy                      1.9.0                      py_0  
readline                  8.0                  h7b6447c_0  
requests                  2.25.0                   pypi_0    pypi
ruamel-yaml               0.16.12                  pypi_0    pypi
ruamel-yaml-clib          0.2.2                    pypi_0    pypi
scikit-image              0.17.2                   pypi_0    pypi
scikit-learn              0.23.2                   pypi_0    pypi
scipy                     1.5.2            py37h0b6359f_0  
send2trash                1.5.0                    py37_0  
setuptools                50.3.2           py37h06a4308_2  
shapely                   1.7.1            py37h98ec03d_0  
sip                       4.19.8           py37hf484d3e_0  
six                       1.15.0           py37h06a4308_0  
sqlite                    3.33.0               h62c20be_0  
statsmodels               0.12.1                   pypi_0    pypi
tables                    3.6.1                    pypi_0    pypi
tabulate                  0.8.7                    pypi_0    pypi
tensorboard               1.13.1           py37hf484d3e_0  
tensorflow                1.13.1          gpu_py37hc158e3b_0  
tensorflow-base           1.13.1          gpu_py37h8d69cac_0  
tensorflow-estimator      1.13.0                     py_0  
tensorflow-gpu            1.13.1               h0d30ee6_0  
tensorpack                0.9.8                    pypi_0    pypi
termcolor                 1.1.0                    py37_1  
terminado                 0.9.1                    py37_0  
testpath                  0.4.4                      py_0  
threadpoolctl             2.1.0                    pypi_0    pypi
tifffile                  2020.12.4                pypi_0    pypi
tk                        8.6.10               hbc83047_0  
tornado                   6.1              py37h27cfd23_0  
tqdm                      4.54.1                   pypi_0    pypi
traitlets                 5.0.5                      py_0  
urllib3                   1.26.2                   pypi_0    pypi
wcwidth                   0.2.5                      py_0  
webencodings              0.5.1                    py37_1  
werkzeug                  1.0.1                      py_0  
wheel                     0.36.0             pyhd3eb1b0_0  
widgetsnbextension        3.5.1                    py37_0  
wxpython                  4.0.4            py37hc99224d_0  
xz                        5.2.5                h7b6447c_0  
zeromq                    4.3.3                he6710b0_3  
zipp                      3.4.0              pyhd3eb1b0_0  
zlib                      1.2.11               h7b6447c_3 

```

**Comment by mfkeles** [comment]
@MorterzaAbbaszadeh  please open a separate issue if you think it is an issue... It sounds like it has nothing to do with the memory issue, I am encountering. That said I would try a fresh install of DLC in a brand new environement. Your DLC version is quite old and lots and lots of improvements have been made since then. It is also possible something in your env is broken....

**Comment by mfkeles** [comment]
hmm, i was trying to play with batch sizes just to see if it has an effect (not sure why it would affect the system memory). I encountered a separate error while the memory was accumulating. Interestingly, when I tried `deeplabcut-docker` I got the same exact error and the analysis stopped at the exact same frame. I don't encounter this problem with batch size 4 or 8. Does this suggest some sort of video related problem @jeylau ? :

```Activating extracting of PAFs
2021-10-30 13:14:48.356017: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:48.407094: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:48.408272: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:48.409642: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-10-30 13:14:48.413366: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:48.414484: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:48.415589: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:49.173430: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:49.174213: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:49.174956: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:937] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-10-30 13:14:49.175672: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 21298 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3090, pci bus id: 0000:21:00.0, compute capability: 8.6
Starting to analyze %  /home/grover/Desktop/20210817_Fly14_F_A_5d_8am-08172021175459.avi
Loading  /home/grover/Desktop/20210817_Fly14_F_A_5d_8am-08172021175459.avi
Duration of video [s]:  57599.3 , recorded with  30.0 fps!
Overall # of frames:  1727979  found with (before cropping) frame dimensions:  1100 800
Starting to extract posture from the video(s) with batchsize: 1
  0%|                                                                                                                                         | 0/1727979 [00:00<?, ?it/s]2021-10-30 13:14:51.376646: I tensorflow/stream_executor/cuda/cuda_dnn.cc:369] Loaded cuDNN version 8101
  1%|█▋                                                                                                                        | 23667/1727979 [21:23<26:22:35, 17.95it/s]---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-7fc1fe0d8e2a> in <module>
----> 1 deeplabcut.analyze_videos(config_path,vid_path,batchsize=1,gputouse=0,save_as_csv=True)

~/Desktop/gitDLC/DeepLabCut/deeplabcut/pose_estimation_tensorflow/predict_videos.py in analyze_videos(config, videos, videotype, shuffle, trainingsetindex, gputouse, save_as_csv, destfolder, batchsize, cropping, TFGPUinference, dynamic, modelprefix, robust_nframes, allow_growth)
    287             )
    288             for video in Videos:
--> 289                 AnalyzeMultiAnimalVideo(
    290                     video,
    291                     DLCscorer,

~/Desktop/gitDLC/DeepLabCut/deeplabcut/pose_estimation_tensorflow/predict_multianimal.py in AnalyzeMultiAnimalVideo(video, DLCscorer, trainFraction, cfg, dlc_cfg, sess, inputs, outputs, destfolder, robust_nframes)
     88             )
     89         else:
---> 90             PredicteData, nframes = GetPoseandCostsS(
     91                 cfg, dlc_cfg, sess, inputs, outputs, vid, nframes,
     92             )

~/Desktop/gitDLC/DeepLabCut/deeplabcut/pose_estimation_tensorflow/predict_multianimal.py in GetPoseandCostsS(cfg, dlc_cfg, sess, inputs, outputs, cap, nframes)
    201                 dlc_cfg, np.expand_dims(frame, axis=0), sess, inputs, outputs,
    202             )
--> 203             PredicteData["frame" + str(counter).zfill(strwidth)] = dets[0]
    204         elif counter >= nframes:
    205             break

IndexError: list index out of range

  1%|█▋                                                                                                                        | 23668/1727979 [21:40<26:22:35, 17In [5]: 

```

**Comment by mfkeles** [comment]
I don't get the same error in a different video recorded with same camera same settings, Still runs out of memory.

**Comment by jeylau** [comment]
@mfkeles, I've experimented with pyav (instead of opencv) here https://github.com/jeylau/DeepLabCut/tree/pyav. Feel free to check the branch out (you'll only need `pip install av==8.0.3` and use `analyze_videos` as usual); it may help with the memory leak 🤞

**Comment by mfkeles** [comment]
![image](https://user-images.githubusercontent.com/22876046/140188717-39432ea9-c317-4dfd-a29b-5d9d46af6145.png)

unfortunately, it is still the same. It improved the speed slightly 5-6%.

**Comment by jeylau** [comment]
@mfkeles, I may have found something! I'll keep digging and hopefully update you soon with a fix 😃

**Comment by mfkeles** [comment]
@jeylau ❤️❤️❤️❤️ can't wait!!

**Comment by jeylau** [comment]
@mfkeles, there is some work in progress here https://github.com/jeylau/DeepLabCut/tree/shelve. All you need is to run `analyze_video(..., use_shelve=True); batches of data will be written to disk on the fly to avoid growing a dictionary in memory. It's still experimental though, as the rest of the pipeline would need a couple of adjustments, but I'd love to get your feedback. If it looks promising, I'll pitch it during our DLC dev meeting 😄

**Comment by mfkeles** [comment]
@jeylau you're a life saver! 50k+ frames and going, no change in the memory!!!!! IT WORKS!

**Comment by mfkeles** [comment]
@jeylau with 50% of the video analyzed, it definitely does not cause any memory issues. However, it is generating rather a large file. (1gb per 1% looks like [~17k frames]). It is not an issue for me as storage is not a big deal, I am wondering though if these files would be required once all the analysis is finished or they are just temp?

**Comment by jeylau** [comment]
Do you analyze an unusually large number of animals/keypoints? That file is like the _full.pickle you were used to, it contains all the information required for tracking (but could later be deleted if you're happy with the final output). I'm glad to hear that memory isn't a problem anymore, but I'll see what can be done now for compression (although 60kb per frame does not seem "excessive" at first sight).

**Comment by mfkeles** [comment]
thanks @jeylau ! 1 animal 35 body parts. I can prob reduce the body part number a bit.

**Comment by auesro** [comment]
Hi @jeylau ,

We (with @isabelperezf)  have just encountered this same issue. 
When using `analyze_video` on some videos, running in COLAB, RAM memory gets completely used and the environment restarts by itself.  The problem seems to be with files largers than 200 MB. We have tested the `use_shelve` parameter with the same results...

Do you have any ideas?

Thanks!
